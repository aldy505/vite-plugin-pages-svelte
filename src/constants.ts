export const MODULE_IDS = ['pages-generated-svelte', 'virtual:generated-pages-svelte'];
export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages-svelte';
