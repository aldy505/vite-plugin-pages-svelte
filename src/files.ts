import type { ResolvedOptions } from './types/options';
import type { FileOutput, PageDirOptions } from './types/page';
import { resolve } from 'path';
import { traverse, traverseDir } from './crawler/crawler';
import { containsExtension } from './utils/validate';

function getIgnore(exclude: string[]): string[] {
  return ['node_modules', '.git', '**/__*__/**', ...exclude];
}

/**
 * Resolves the page dirs for its for its given globs
 * @param pageDirOptions
 * @param root
 * @param exclude
 * @returns {Promise<PageDirOptions[]>}
 */
export async function getPageDirs(
  pageDirOptions: PageDirOptions,
  root: string,
  exclude: string[],
): Promise<PageDirOptions[]> {
  const dirs = await traverseDir(resolve(root, pageDirOptions.baseRoute, pageDirOptions.dir), getIgnore(exclude));
  console.log(dirs);
  const pageDirs = dirs.map((dir) => ({
    ...pageDirOptions,
    dir,
  }));

  return pageDirs;
}

/**
 * Resolves the files that are valid pages for the given context.
 * @param path
 * @param options
 * @returns
 */
export async function getPageFiles(path: string, options: ResolvedOptions): Promise<FileOutput[]> {
  const { exclude, extensions } = options;

  const files = await traverse(path, extensions, exclude);

  return files;
}

/**
 * Generate a FileOutput object from the given path and options.
 * @param path
 * @param options
 * @returns
 */
export function fromSinglePage(path: string, options: ResolvedOptions): FileOutput {
  const { exclude, extensions } = options;
  if (exclude.includes(path) && extensions.length > 0 && !containsExtension(path, extensions)) return { path: '' };
  return { path };
}
