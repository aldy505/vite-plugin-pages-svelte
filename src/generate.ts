import type { PreRoute } from './types/route';
import type { FileOutput } from './types/page';
import { isDynamicRoute, isCatchAllRoute } from './utils/validate';
import { stringifyRoutes } from './stringify';
import { haveChildren } from './crawler/crawler';

export function generateRoutes(pages: FileOutput[]): PreRoute[] {
  const routes: PreRoute[] = [];

  for (let i = 0; i < pages.length; i++) {
    const node = pages[i].path;
    const isDynamic = isDynamicRoute(node);
    const isCatchAll = isCatchAllRoute(node);
    const normalizedName = isDynamic ? node.replace(/^\[(\.{3})?/, '').replace(/\]$/, '') : node;
    const normalizedPath = normalizedName.toLowerCase();
    let name: string;
    if (normalizedName.toLowerCase() === 'index' && !node) {
      name = 'index';
    } else {
      if (isCatchAll) {
        name = '/*';
      } else if (isDynamic) {
        name = `/:${normalizedName}`;
      } else {
        name = `/${normalizedPath}`;
      }
    }

    if (!haveChildren(pages[i])) {
      routes.push({
        name,
        path: pages[i].path,
      });
      continue;
    }
    routes.push({
      name,
      path: pages[i].path,
      children: generateRoutes(pages[i].children as FileOutput[]),
    });
  }

  return routes;
}

export function generateClientCode(routes: PreRoute[]): string {
  const { imports, stringRoutes } = stringifyRoutes(routes);

  return (
    `import {wrap} from 'svelte-spa-router/wrap';\n` +
    `${imports.join(';\n')}${imports.length > 1 ? ';' : ''}\n\n` +
    `const routes = ${stringRoutes};\n\n` +
    `export default routes;`
  );
}
