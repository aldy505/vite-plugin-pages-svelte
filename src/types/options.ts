/**
 * Plugin options.
 */
interface Options {
  /**
   * Relative path to the directory to search for page components.
   * @default 'src/pages'
   */
  pagesDir: string;
  /**
   * Valid file extensions for page components.
   * @default ['svelte']
   */
  extensions: string[];
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[];
  /**
   * Sync load top level index file
   * @default true
   */
  syncIndex: boolean;
}

export type UserOptions = Partial<Options>;

export interface ResolvedOptions extends Options {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string;
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp;
}

export type ImportMode = 'sync' | 'async';
export type ImportModeResolveFn = (filepath: string) => ImportMode;
