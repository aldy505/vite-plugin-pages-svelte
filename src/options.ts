import type { UserOptions, ResolvedOptions } from './types/options';
import type { PageDirOptions } from './types/page';
import { getPageDirs } from './files';
import { slash } from './utils/convert';

/**
 *
 * @param pagesDir
 * @param root
 * @param exclude
 * @returns
 */
async function resolvePageDirs(pagesDir: string, root: string, exclude: string[]): Promise<PageDirOptions[]> {
  return await getPageDirs({ baseRoute: '', dir: pagesDir }, root, exclude);
}

/**
 *
 * @param userOptions
 * @param viteRoot
 * @returns
 */
export async function resolveOptions(userOptions: UserOptions, viteRoot?: string): Promise<ResolvedOptions> {
  const { pagesDir = 'src/pages', exclude = [], syncIndex = true } = userOptions;

  const root = viteRoot || slash(process.cwd());

  const extensions = userOptions.extensions || ['svelte'];

  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`);

  const resolvedPagesDir = await resolvePageDirs(slash(pagesDir), root, exclude);

  const resolvedOptions: ResolvedOptions = {
    pagesDir: resolvedPagesDir[0].dir,
    root,
    extensions,
    exclude,
    syncIndex,
    extensionsRE,
  };

  return resolvedOptions;
}
