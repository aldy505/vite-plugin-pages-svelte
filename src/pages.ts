import type { ResolvedOptions } from './types/options';
import type { FileOutput } from './types/page';
import { resolve } from 'path';
import { getPageFiles } from './files';
import { toArray, slash } from './utils/convert';

export async function resolvePages(options: ResolvedOptions): Promise<FileOutput[]> {
  const dirs = toArray(options.pagesDir);

  const pages: FileOutput[] = [];

  const pageDirFiles = dirs.map(async (pageDir) => {
    const pagePath = slash(resolve(options.root, pageDir));
    return {
      pageDir,
      files: await getPageFiles(pagePath, options),
    };
  });

  for await (const pageDir of pageDirFiles) {
    for (const file of pageDir.files) {
      pages.push(file);
    }
  }

  const routes: string[] = [];

  for (const page of pages.values()) {
    if (!routes.includes(page.path)) routes.push(page.path);
    else throw new Error(`[vite-plugin-pages] duplicate route in ${page.path}`);
  }

  return pages;
}

export function addPage(pages: FileOutput[], newPage: FileOutput): void {
  pages.push(newPage);
}

export function removePage(pages: FileOutput[], oldPage: FileOutput): void {
  pages = pages.filter((o) => o.path !== oldPage.path);
}
