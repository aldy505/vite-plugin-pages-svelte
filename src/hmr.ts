import type { ViteDevServer } from 'vite';
import type { ResolvedOptions } from './types/options';
import type { FileOutput } from './types/page';
import { isTarget } from './utils/validate';
import { slash } from './utils/convert';
import { getPagesVirtualModule, debug } from './utils/vite';
import { addPage, removePage } from './pages';
import { fromSinglePage } from './files';

export function handleHMR(
  server: ViteDevServer,
  pages: FileOutput[],
  options: ResolvedOptions,
  clearRoutes: () => void,
): void {
  const { ws, watcher } = server;

  function fullReload() {
    // invalidate module
    getPagesVirtualModule(server);
    clearRoutes();
    ws.send({
      type: 'full-reload',
    });
  }

  watcher.on('add', async (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      const pageFile = fromSinglePage(path, options);
      addPage(pages, pageFile);
      debug.hmr('add', path);
      fullReload();
    }
  });
  watcher.on('unlink', (file) => {
    const path = slash(file);
    if (isTarget(path, options)) {
      const pageFile = fromSinglePage(path, options);
      removePage(pages, pageFile);
      debug.hmr('remove', path);
      fullReload();
    }
  });
}
