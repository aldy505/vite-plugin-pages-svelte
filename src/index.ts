import type { Plugin } from 'vite';
import type { ResolvedOptions, UserOptions } from './types/options';
import type { PreRoute } from './types/route';
import type { FileOutput } from './types/page';
import { generateRoutes, generateClientCode } from './generate';
import { debug } from './utils/vite';
import { resolveOptions } from './options';
import { MODULE_IDS, MODULE_ID_VIRTUAL } from './constants';
import { resolvePages } from './pages';
import { handleHMR } from './hmr';

function pagesPlugin(userOptions: UserOptions = {}): Plugin {
  let generatedRoutes: PreRoute[] | null = null;
  let options: ResolvedOptions;
  let pages: FileOutput[];

  return {
    name: 'vite-plugin-pages-svelte',
    enforce: 'pre',
    async configResolved({ root }) {
      options = await resolveOptions(userOptions, root);
      pages = await resolvePages(options);
      debug.options(options);
      debug.pages(pages);
    },
    configureServer(server) {
      handleHMR(server, pages, options, () => {
        generatedRoutes = null;
      });
    },
    resolveId(id) {
      return MODULE_IDS.includes(id) || MODULE_IDS.some((i) => id.startsWith(i)) ? MODULE_ID_VIRTUAL : null;
    },
    load(id) {
      if (id !== MODULE_ID_VIRTUAL) return;

      if (!generatedRoutes) {
        generatedRoutes = generateRoutes(pages);
      }

      const clientCode = generateClientCode(generatedRoutes);
      debug.gen('client code: %O', clientCode);

      return clientCode;
    },
    transform(_code, id) {
      if (!/svelte&type=route/.test(id)) return;
      return {
        code: 'export default {};',
        map: null,
      };
    },
    generateBundle(_options, bundle) {
      bundle;
    },
  };
}

export * from './types/options';
export { generateRoutes };
export default pagesPlugin;
