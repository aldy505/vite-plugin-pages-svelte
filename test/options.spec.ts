import { resolveOptions } from '../src/options';

describe('Options', () => {
  test('resolve', async () => {
    const options = await resolveOptions({
      pagesDir: 'test/assets/pages',
    });
    expect(options).toMatchSnapshot(
      {
        root: expect.any(String),
      },
      'resolved options',
    );
  });
});
