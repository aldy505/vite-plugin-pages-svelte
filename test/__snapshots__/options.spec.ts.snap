// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options resolve: resolved options 1`] = `
Object {
  "exclude": Array [],
  "extensions": Array [
    "svelte",
  ],
  "extensionsRE": /\\\\\\.\\(svelte\\)\\$/,
  "pagesDir": Array [
    Object {
      "baseRoute": "",
      "dir": "test/assets/pages",
    },
  ],
  "root": Any<String>,
  "syncIndex": true,
}
`;
